libraries/chain/block_state.cpp

```diff
@@ -233,7 +238,14 @@ vote_status_t block_state::has_voted(const bls_public_key& key) const {
 
 // Called from net threads
 void block_state::verify_qc(const qc_t& qc) const {
-   aggregating_qc.verify_qc(qc, strong_digest, weak_digest);
+   // Do not use `block_state::aggregating_qc` which applies only for `this` block.
+   // `verify_qc()` can be called on a descendant `block_state` of `qc.block_num`, so we need
+   // to create a new `aggregating_qc_t` with the finalizer policies of the claimed block.
+   // ---------------------------------------------------------------------------------------
+   finalizer_policies_t policies = get_finalizer_policies(qc.block_num);
+   aggregating_qc_t aggregating_qc(policies.active_finalizer_policy, policies.pending_finalizer_policy);
+
+   aggregating_qc.verify_qc(qc, policies.finality_digest, create_weak_digest(policies.finality_digest));
```

Comment:
With this usage, probably makes sense to factor out qc verification out of `aggregating_qc_t` into another type that `aggregating_qc_t` and this can use. The `active_policy_sig` and `pending_policy_sig` of `aggregating_qc_t` are not used. Seems wasteful to create a `aggregating_qc_t` for this.

Maybe add a `finalizer_policy_pair_t` that both `aggregating_qc_t` and this method can use. The `finalizer_policy_pair_t` can have the `verify_qc` and `verify_dual_finalizers_votes` methods.

Reply:
Instead of `block_state` having a `verify_qc`, maybe it makes more sense for it to return a `finalizer_policy_pair_t` given a `block_num`.

Reply:
I don't disagree, but I think we should refrain from refactoring the code in a 1.01 release. 
I tried to do what was necessary for cleany fixing the issue in this PR.
I'll be happy to clean up this further in a separate PR targeted to the main branch.

Reply:
Please create an issue for it if not doing it in this PR. It is wasteful as is.